File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/keyguard/SystemUIChargeWattsHook.java
package com.qimian233.ztool.hook.modules.systemui.keyguard;

import android.annotation.SuppressLint;

import com.qimian233.ztool.hook.base.BaseHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Method;

/**
 * SystemUI充电瓦数显示Hook模块
 * 在锁屏充电提示中添加实时充电功率显示
 */
@SuppressLint("PrivateApi")
public class SystemUIChargeWattsHook extends BaseHookModule {

    private static final String TARGET_CLASS = "com.android.systemui.statusbar.KeyguardIndicationController";

    public SystemUIChargeWattsHook() {}

    @Override
    public String getModuleName() {
        return "systemui_charge_watts";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                "com.android.systemui"
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if ("com.android.systemui".equals(packageName)) {
            hookKeyguardIndicationController(classLoader);
        }
    }

    private void hookKeyguardIndicationController(ClassLoader classLoader) {
        try {
            // Hook computePowerIndication方法来添加充电瓦数显示
            Method computeMethod = classLoader.loadClass(TARGET_CLASS)
                    .getDeclaredMethod("computePowerIndication");
            hookWithId(computeMethod, "compute", chain -> {
                try {
                    // 获取原始返回的充电提示文本
                    Object result = chain.proceed();
                    String originalText = (String) result;
                    if (originalText == null) return null;

                    // 获取KeyguardIndicationController实例
                    Object controller = chain.getThisObject();
                    Class<?> cl = controller.getClass();

                    // 获取充电状态相关字段
                    boolean isPluggedIn = cl.getDeclaredField("mPowerPluggedIn").getBoolean(controller);
                    int chargingWattage = cl.getDeclaredField("mChargingWattage").getInt(controller);
                    int chargingSpeed = cl.getDeclaredField("mChargingSpeed").getInt(controller);

                    // 只在充电状态下显示瓦数，且瓦数大于0
                    if (isPluggedIn && chargingWattage > 0) {
                        // 尝试多种单位转换
                        int watts = calculateActualWatts(chargingWattage);

                        if (watts > 0) {
                            // 使用换行符 \n 追加功率信息
                            String newText = originalText + "\n" + formatWattage(watts, chargingSpeed);
                            logger.debug("成功添加充电瓦数显示: " + watts + "W, speed=" + chargingSpeed);
                            return newText;
                        }
                    }
                    return result;
                } catch (Throwable t) {
                    logger.error("computePowerIndication hook回调异常", t);
                    return chain.proceed();
                }
            });

            // 额外Hook电池状态更新方法，确保能获取到最新的充电数据
            // onRefreshBatteryInfo 在新版 SystemUI 中位于内部类 BaseKeyguardCallback 中
            Method refreshMethod = null;
            try {
                Class<?> callbackClass = classLoader.loadClass(
                        "com.android.systemui.statusbar.KeyguardIndicationController$BaseKeyguardCallback");
                Class<?> batteryStatusClass = classLoader.loadClass(
                        "com.android.settingslib.fuelgauge.BatteryStatus");
                refreshMethod = callbackClass.getDeclaredMethod("onRefreshBatteryInfo", batteryStatusClass);
            } catch (NoSuchMethodException | ClassNotFoundException e) {
                logger.warn("Unable to find BaseKeyguardCallback.onRefreshBatteryInfo: " + e.getMessage());
            }
            if (refreshMethod != null) {
                hookWithId(refreshMethod, "final_refresh", chain -> {
                    try {
                        Object result = chain.proceed();
                        // 这个方法会在电池状态更新时调用，我们可以在这里获取最新的充电数据
                        Object batteryStatus = chain.getArg(0);
                        if (batteryStatus != null) {
                            try {
                                // 尝试从BatteryStatus对象获取充电功率
                                int maxChargingWattage = batteryStatus.getClass()
                                        .getDeclaredField("maxChargingWattage").getInt(batteryStatus);
                                // BaseKeyguardCallback 是 KeyguardIndicationController 的非静态内部类
                                // 通过 this$0 获取外部类实例
                                Object callback = chain.getThisObject();
                                java.lang.reflect.Field outerField = callback.getClass()
                                        .getDeclaredField("this$0");
                                outerField.setAccessible(true);
                                Object controller = outerField.get(callback);
                                Class<?> cl = controller.getClass();

                                // 记录调试信息
                                logger.debug("BatteryStatus更新 - maxChargingWattage: " + maxChargingWattage +
                                        ", mChargingWattage: " + cl.getDeclaredField("mChargingWattage").getInt(controller));

                            } catch (Throwable t) {
                                logger.error("读取BatteryStatus失败", t);
                            }
                        }
                        return result;
                    } catch (Throwable t) {
                        logger.error("onRefreshBatteryInfo hook回调异常", t);
                        return chain.proceed();
                    }
                });
            } else {
                logger.warn("Cannot find onRefreshBatteryInfo, skipping this hook");
            }

            logger.info("成功Hook KeyguardIndicationController");

        } catch (Throwable t) {
            logger.error("Hook KeyguardIndicationController失败", t);
        }
    }

    /**
     * 尝试多种方式计算实际瓦数
     */
    private int calculateActualWatts(int rawWattage) {
        // 情况1：如果值在合理范围内（1-150W），直接使用
        if (rawWattage > 0 && rawWattage <= 150000) {
            // 可能是毫瓦单位，转换为瓦
            return rawWattage / 1000;
        }

        // 情况2：如果值很大，可能是微瓦单位
        if (rawWattage > 150000 && rawWattage <= 150000000) {
            return rawWattage / 1000000;
        }

        // 情况3：如果值很小，可能是直接就是瓦数
        if (rawWattage > 0 && rawWattage <= 150) {
            return rawWattage;
        }

        // 情况4：如果值异常大，尝试除以10000（某些设备的特殊单位）
        if (rawWattage > 1000000) {
            return rawWattage / 10000;
        }

        // 无法确定单位，返回0表示不显示
        logger.warn("无法识别的瓦数单位: " + rawWattage);
        return 0;
    }

    /**
     * 格式化充电瓦数显示：显示"[功率]W [闪电符号]"
     * 根据 mChargingSpeed 字段判断充电速度等级并附加闪电符号
     * @param watts 充电功率（瓦）
     * @param chargingSpeed 充电速度等级：1=慢速, 2=快速, 3=极速
     */
    private String formatWattage(int watts, int chargingSpeed) {
        if (watts <= 0) return "";

        // 基础字符串："[功率]W"
        String base = watts + "W";

        // 根据充电速度等级附加闪电符号
        switch (chargingSpeed) {
            case 3:
                return base + "⚡⚡";  // 极速充电
            case 2:
                return base + "⚡";    // 快速充电
            case 1:
            default:
                return base;              // 慢速充电，无闪电符号
        }
    }

}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/keyguard/SystemUIChargeWattsHook.kt
package com.qimian233.ztool.hook.modules.systemui.keyguard

import android.annotation.SuppressLint
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Method

/**
 * SystemUI charging wattage display hook module.
 * Adds a real-time charging power display to the lock screen charging indication.
 */
@SuppressLint("PrivateApi")
class SystemUIChargeWattsHook : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.SYSTEMUI_CHARGE_WATTS.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.SYSTEM_UI.packageName)

    @Throws(Throwable::class)
    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        hookKeyguardIndicationController(classLoader)
    }

    private fun hookKeyguardIndicationController(classLoader: ClassLoader) {
        try {
            // Hook computePowerIndication to add the charging wattage display
            val computeMethod = classLoader.loadClass(TARGET_CLASS)
                .getDeclaredMethod("computePowerIndication")
            hookWithId(computeMethod, "compute") { chain ->
                try {
                    // Get the original charging indication text
                    val result = chain.proceed()
                    val originalText = result as String

                    // Get the KeyguardIndicationController instance
                    val controller = chain.thisObject
                    val cl: Class<*> = controller.javaClass

                    // Get charging-state related fields
                    val isPluggedIn = cl.getDeclaredField("mPowerPluggedIn").getBoolean(controller)
                    val chargingWattage = cl.getDeclaredField("mChargingWattage").getInt(controller)
                    val chargingSpeed = cl.getDeclaredField("mChargingSpeed").getInt(controller)

                    // Only show wattage while charging and when wattage is greater than 0
                    if (isPluggedIn && chargingWattage > 0) {
                        // Try multiple unit conversions
                        val watts = calculateActualWatts(chargingWattage)

                        if (watts > 0) {
                            // Append power info using newline \n
                            val newText = originalText + "\n" + formatWattage(watts, chargingSpeed)
                            logger.debug("Charging wattage display added: " + watts + "W, speed=" + chargingSpeed)
                            return@hookWithId newText
                        }
                    }
                    return@hookWithId result
                } catch (t: Throwable) {
                    logger.error("computePowerIndication hook callback error", t)
                    return@hookWithId chain.proceed()
                }
            }

            // Additionally hook the battery status update method to get the latest charging data
            // onRefreshBatteryInfo lives in the inner class BaseKeyguardCallback on newer SystemUI versions
            var refreshMethod: Method? = null
            try {
                val callbackClass = classLoader.loadClass(
                    $$"com.android.systemui.statusbar.KeyguardIndicationController$BaseKeyguardCallback"
                )
                val batteryStatusClass = classLoader.loadClass(
                    "com.android.settingslib.fuelgauge.BatteryStatus"
                )
                refreshMethod =
                    callbackClass.getDeclaredMethod("onRefreshBatteryInfo", batteryStatusClass)
            } catch (e: NoSuchMethodException) {
                logger.warn("Unable to find BaseKeyguardCallback.onRefreshBatteryInfo: " + e.message)
            } catch (e: ClassNotFoundException) {
                logger.warn("Unable to find BaseKeyguardCallback.onRefreshBatteryInfo: " + e.message)
            }
            if (refreshMethod != null) {
                hookWithId(refreshMethod, "final_refresh") { chain ->
                    try {
                        val result = chain.proceed()
                        // This method is called on battery status updates, so we can get the latest charging data here
                        val batteryStatus = chain.args[0]
                        if (batteryStatus != null) {
                            try {
                                // Try to get the charging power from the BatteryStatus object
                                val maxChargingWattage = batteryStatus.javaClass
                                    .getDeclaredField("maxChargingWattage").getInt(batteryStatus)
                                // BaseKeyguardCallback is a non-static inner class of KeyguardIndicationController;
                                // get the outer class instance via this$0
                                val callback = chain.thisObject
                                val outerField = callback.javaClass
                                    .getDeclaredField("this$0")
                                outerField.isAccessible = true
                                val controller = outerField.get(callback)
                                val cl: Class<*> = controller!!.javaClass

                                // Log debug info
                                logger.debug(
                                    "BatteryStatus update - maxChargingWattage: " + maxChargingWattage +
                                            ", mChargingWattage: " + cl.getDeclaredField("mChargingWattage")
                                        .getInt(controller)
                                )
                            } catch (t: Throwable) {
                                logger.error("Failed to read BatteryStatus", t)
                            }
                        }
                        return@hookWithId result
                    } catch (t: Throwable) {
                        logger.error("onRefreshBatteryInfo hook callback error", t)
                        return@hookWithId chain.proceed()
                    }
                }
            } else {
                logger.warn("Cannot find onRefreshBatteryInfo, skipping this hook")
            }

            logger.info("Successfully hooked KeyguardIndicationController")
        } catch (t: Throwable) {
            logger.error("Failed to hook KeyguardIndicationController", t)
        }
    }

    /**
     * Try multiple ways to calculate the actual wattage
     */
    private fun calculateActualWatts(rawWattage: Int): Int {
        // Case 1: value in a plausible range (1-150W), use directly
        if (rawWattage in 1..150000) {
            // Likely in milliwatts, convert to watts
            return rawWattage / 1000
        }

        // Case 2: very large value, likely in microwatts
        if (rawWattage in 150001..150000000) {
            return rawWattage / 1000000
        }

        // Case 3: abnormally large value, try dividing by 10000 (device-specific unit)
        if (rawWattage > 1000000) {
            return rawWattage / 10000
        }

        // Unknown unit, return 0 to hide the display
        logger.warn("Unrecognized wattage unit: $rawWattage")
        return 0
    }

    /**
     * Format the charging wattage display: "<power>W <lightning symbols>"
     * Appends lightning symbols based on the mChargingSpeed level.
     * @param watts charging power (watts)
     * @param chargingSpeed charging speed level: 1=slow, 2=fast, 3=turbo
     */
    private fun formatWattage(watts: Int, chargingSpeed: Int): String {
        if (watts <= 0) return ""

        // Base string: "[power]W"
        val base = watts.toString() + "W"

        // Append lightning symbols based on the charging speed level
        return when (chargingSpeed) {
            3 -> "$base⚡⚡" // Turbo charging
            2 -> "$base⚡" // Fast charging
            1 -> base // Slow charging, no lightning symbol
            else -> base
        }
    }

    companion object {
        private const val TARGET_CLASS =
            "com.android.systemui.statusbar.KeyguardIndicationController"
    }
}
